components/utilities/ai.js

```diff
@@ -0,0 +1,184 @@
+/* AI logic to play a double */
+// Need to handle logic if there are 3/4 of a kind
+// This leads to different combinations of doubles.
+const double = (hand) => {
+  const dupesObject = createDuplicatesObject(hand);
+  const combinations = [];
+
+  for (const value in dupesObject) {
+    if (dupesObject[value].length === 2) {
+      combinations.push(dupesObject[value]);
+    } else if (dupesObject[value].length > 2) {
+      for (let i = 0; i < dupesObject[value].length; i++) {
+        for (let j = i + 1; j < dupesObject[value].length; j++) {
+          combinations.push([dupesObject[value][i], dupesObject[value][j]]);
+        }
+      }
+    }
+  }
+  return combinations;
+}
+/* AI logic to play a triple */
+/** 
+ * NOTE: This solution for if the number of duplicates is four seems severely need of refactoring.
+*/
+const triple = (hand) => {
+  const dupesObject = createDuplicatesObject(hand);
+  const combinations = [];
+
+  for (const value in dupesObject) {
+    if (dupesObject[value].length === 3) {
+      combinations.push(dupesObject[value]);
+    } else if (dupesObject[value].length === 4) {
+      for (let i = 0; i < dupesObject[value].length; i++) {
+        for (let j = i + 1; j < dupesObject[value].length; j++) {
+          if (dupesObject[value][j + 1]) {
+            combinations.push([dupesObject[value][i], dupesObject[value][j], dupesObject[value][j + 1]]);
+          }
+
+          if (dupesObject[value][j + 2]) {
+            combinations.push([dupesObject[value][i], dupesObject[value][j], dupesObject[value][j + 2]]);
+          }
+        }
+      }
+    }
+  }
+  return combinations;
+}
+/* AI logic to play a quartet */
+const quartet = (hand) => {
+  const dupesObject = createDuplicatesObject(hand);
+  const combinations = [];
+
+  for (const value in dupesObject) {
+    if (dupesObject[value].length === 4) {
+      combinations.push(dupesObject[value]);
+    }
+  }
+  return combinations;
+}
+
+/* AI logic to play double sequence */
+
+
+const merge = (left, right) => {
+  let sortedArr = [];
+
+  while (left.length && right.length) {
```

Comment:
hmm, am i missing something? in theory `left` and `right` could have a length, thus this `while` loop will never break?

Reply:
In the while loop, the array is being reduced until it is empty by using left.shift() and right.shift()